doc/api/stream.md

```diff
@@ -90,60 +97,224 @@ server.listen(1337);
+* [fs write streams][]
 * [zlib streams][zlib]
 * [crypto streams][crypto]
+* [TCP sockets][]
+* [child process stdin][]
+* [`process.stdout`][], [`process.stderr`][]
 
-### Class: stream.Readable
+All [Writable][] streams implement the interface defined by the
+`stream.Writable` class.
+
+While specific instances of [Writable][] streams may differ in various ways,
+all Writable streams follow the same fundamental usage pattern as illustrated
+in the example below:
+
+```js
+const myStream = getWritableStreamSomehow();
+myStream.write('some data');
+myStream.write('some more data');
+myStream.end('done writing data');
+```
+
+#### Class: stream.Writable
 
 <!--type=class-->
 
-The Readable stream interface is the abstraction for a *source* of
-data that you are reading from. In other words, data comes *out* of a
-Readable stream.
+##### Event: 'close'
 
-A Readable stream will not start emitting data until you indicate that
-you are ready to receive it.
+The `'close'` event is emitted when the stream and any of its underlying
+resources (a file descriptor, for example) have been closed. The event indicates
+that no more events will be emitted, and no further computation will occur.
 
-Readable streams have two "modes": a **flowing mode** and a **paused
-mode**. When in flowing mode, data is read from the underlying system
-and provided to your program as fast as possible. In paused mode, you
-must explicitly call [`stream.read()`][stream-read] to get chunks of data out.
-Streams start out in paused mode.
+Not all Writable streams will emit the `'close'` event.
 
-**Note**: If no data event handlers are attached, and there are no
-[`stream.pipe()`][] destinations, and the stream is switched into flowing
-mode, then data will be lost.
+##### Event: 'drain'
 
-You can switch to flowing mode by doing any of the following:
+If a call to [`stream.write(chunk)`][stream-write] returns `false`, the
+`'drain'` event will be emitted when it is appropriate to resume writing data
+to the stream.
 
-* Adding a [`'data'`][] event handler to listen for data.
-* Calling the [`stream.resume()`][stream-resume] method to explicitly open the
-  flow.
-* Calling the [`stream.pipe()`][] method to send the data to a [Writable][].
+```js
+// Write the data to the supplied writable stream one million times.
+// Be attentive to back-pressure.
+function writeOneMillionTimes(writer, data, encoding, callback) {
+  var i = 1000000;
+  write();
+  function write() {
+    var ok = true;
+    do {
+      i--;
+      if (i === 0) {
+        // last time!
+        writer.write(data, encoding, callback);
+      } else {
+        // see if we should continue, or wait
+        // don't pass the callback, because we're not done yet.
+        ok = writer.write(data, encoding);
+      }
+    } while (i > 0 && ok);
+    if (i > 0) {
+      // had to stop early!
+      // write some more once it drains
+      writer.once('drain', write);
+    }
+  }
+}
+```
```

Comment:
> In my experience no, this is not correct. There is a cost in having a huge stream buffer: you have a bunch of data D, which you convert to C and then write into your stream. This allocates size(D) + size(C) at the beginning, and then it reduces it. If you write chunk by chunk, you have size(D) + size(highWaterMark * 2) at the beginning.
> Even if you write strings these are converted to Buffers internally in most cases. You can probably write a case were this does not hold, but it is not the default.

OTOH the example above will write synchronously until HWM, then write one chunk for every chunk flushed asynchronously — vastly slowing down the process. It would seem to be a tradeoff between time and space.

> Plus, most API can be written to support backpressure: it is just a design habit. If you design an API to support backpressure, you can always ignore it. The reverse is not possible, as backpressure is usually something costly to add later.

Indeed — I agree here — but I believe it's important to give examples of common situations, and it's fairly common to encounter cases where you may have all of the data on-hand, synchronously available, and you need to push it through a (hopefully `writev`-backed!) Writable stream. The answer seems to be "trade code complexity & time cost for memory efficiency" _or_ "trade memory efficiency for faster wall-clock time (but potentially being harsher to the event loop & memory.)" Maybe it's important to note both approaches? Or note the reasoning behind the former? 

In general, I'm not a huge fan of encouraging folks to implement backpressure using `.write`, since `.pipe` already does that & it is intricate enough that it is easy to get wrong.